src/main/java/seedu/address/storage/JsonAdaptedApplication.java

```diff
@@ -0,0 +1,104 @@
+     */
+    @JsonCreator
+    public JsonAdaptedApplication(@JsonProperty("company") String company, @JsonProperty("contact") String contact,
+                                  @JsonProperty("email") String email, @JsonProperty("position") String position,
+                                  @JsonProperty("date") LocalDate date) {
+        this.company = company;
+        this.contact = contact;
+        this.email = email;
+        this.position = position;
+        this.date = date;
+    }
+
+    /**
+     * Converts a given {@code Application} into this class for Jackson use.
+     */
+    public JsonAdaptedApplication(Application source) {
+        company = source.getCompany().company;
+        contact = source.getContact().value;
+        email = source.getEmail().value;
+        position = source.getPosition().value;
+        date = source.getDate().value;
+    }
+
+    /**
+     * Converts this Jackson-friendly adapted application object into the model's {@code Application} object.
+     *
+     * @throws IllegalValueException if there were any data constraints violated in the adapted application.
+     */
+    public Application toModelType() throws IllegalValueException {
+        if (company == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, Company.class.getSimpleName()));
+        }
+        if (!Company.isValidCompany(company)) {
+            throw new IllegalValueException(Company.MESSAGE_CONSTRAINTS);
+        }
+        final Company modelCompany = new Company(company);
+
+        if (contact == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, Contact.class.getSimpleName()));
+        }
+        if (!Contact.isValidContact(contact)) {
+            throw new IllegalValueException(Contact.MESSAGE_CONSTRAINTS);
+        }
+        final Contact modelContact = new Contact(contact);
+
+        if (email == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, Email.class.getSimpleName()));
+        }
+        if (!Email.isValidEmail(email)) {
+            throw new IllegalValueException(Email.MESSAGE_CONSTRAINTS);
+        }
+        final Email modelEmail = new Email(email);
+
+        if (position == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT,
+                    Position.class.getSimpleName()));
+        }
+        if (!Position.isValidPosition(position)) {
+            throw new IllegalValueException(Position.MESSAGE_CONSTRAINTS);
+        }
+        final Position modelPosition = new Position(position);
+
+        if (date == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT, Date.class.getSimpleName()));
+        }
+        if (!Date.isValidDate(date.toString())) {
+            throw new IllegalValueException(Date.MESSAGE_CONSTRAINTS);
+        }
```

Comment:
While writing the test, I notice this condition is slightly not checkable. Since date is a `LocalDate`, `date.toString()` will always be valid because in `application.Date` class, we validate the String according to `LocalDate.toString()` format? 

I'm thinking of changing it to using string type as the attribute, any other concern from your side? 😅

Reply:
Actually if `date` is already a `LocalDate`, then I think it's guaranteed to be valid as long as it's not null. If that's the case, maybe we can just do a null check and then save the `LocalDate` into our `Date` object?

Reply:
Oh because in storage test, we can test if the json (probably can check if the json file is modified manually before?) file stores invalid date String format (eg.`23-01-2022`) and throw the `IllegalValueException`. See [link](https://github.com/AY2223S1-CS2103-F14-3/tp/pull/59/files#diff-5142b0da70807a19ac9dbe9cb7da1fb3c97cab289e2e69a10c30b94afdc6caa3)